### PR TITLE
feat(storage): enable pause and cancel APIs for download operations

### PR DESCRIPTION
**Pull request by HuiSF**
*Issue #, if available:*

*Description of changes:*

Functionality of canceling a download operation depends on https://github.com/aws-amplify/amplify-flutter/pull/2545.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/storage/amplify_storage_s3_dart/test/storage_s3_service/task/s3_download_task_test.dart [hunk 1]
```diff
@@ -266,6 +271,7 @@ void main() {
         await downloadTask.pause();
```
Comment (dnys1) [inline packages/storage/amplify_storage_s3_dart/test/storage_s3_service/task/s3_download_task_test.dart:271]: Why is `start` not awaited on the line above?

**Review by dnys1 — APPROVED** [review]

File: packages/storage/amplify_storage_s3_dart/test/storage_s3_service/task/s3_download_task_test.dart (hunk 1 above)
Comment (HuiSF) [reply packages/storage/amplify_storage_s3_dart/test/storage_s3_service/task/s3_download_task_test.dart:271]: `task.start` is an async function that creates async actions internally, e.g.

* prefix resolution
* Getting response from the `getObject` call


It's called when creating a download operation, and so is `unawaited`, as it should not block returning the created operation. The final state of `task.start` is determined by `await task.result.`


`task.pause` internally awaits the async action (if any) that is created by `task.start.` Using `unawaited` (task.start)  here replicates the actual flow in the implementation.

Comment (dnys1) [reply packages/storage/amplify_storage_s3_dart/test/storage_s3_service/task/s3_download_task_test.dart:271]: Isn't it more common for a user to call `pause` sometime after `start` has completed?

Comment (HuiSF) [reply packages/storage/amplify_storage_s3_dart/test/storage_s3_service/task/s3_download_task_test.dart:271]: A user is not aware of the existence of `task.start`, even the task itself is hidden and wrapped by the operation.

```dart
final operation = Amplify.Storage.downloadData(...); // creates a _task, and call _task.start()

// a developer can call operation.pause() any time after getting the operation reference 
operation.pause(); // calls _task.pause() and
                   // the pause logic starts executing only when the async
                   // actions created by _task.start() complete.
```

**Review by ragingsquirrel3 — APPROVED** [review]

File: packages/storage/amplify_storage_s3_dart/test/storage_s3_service/task/s3_download_task_test.dart (hunk 1 above)
Comment (dnys1) [reply packages/storage/amplify_storage_s3_dart/test/storage_s3_service/task/s3_download_task_test.dart:271]: Okay. Just wondering if waiting for the task to start makes a difference then? Ideally, it shouldn't right?

Comment (HuiSF) [reply packages/storage/amplify_storage_s3_dart/test/storage_s3_service/task/s3_download_task_test.dart:271]: Correct, it doesn't. The behaviors of pausing on the following

```dart
unawaited(downloadTask.start());
await downloadTask.pause();
// OR
await downloadTask.start();
await downloadTask.pause();
```
are identical.

Comment (dnys1) [reply packages/storage/amplify_storage_s3_dart/test/storage_s3_service/task/s3_download_task_test.dart:271]: Cool - can we change it to the latter then? 😅 `unawaited` makes it seem like you're specifically trying to pause before start finishes.

Comment (HuiSF) [reply packages/storage/amplify_storage_s3_dart/test/storage_s3_service/task/s3_download_task_test.dart:271]: I'd like to keep it as is, as this is how it's being called when creating the operation :D will add a comment to clarify 👍

**Pull request merged**